### ISSUE TITLE
Rename Which -> Whose

### ISSUE THREAD

**Issue by cmeeren** [fluentassertions/fluentassertions#577]
"Whose" is the possessive of the relative "which". See [Wikipedia: English possessive](https://en.wikipedia.org/wiki/English_possessive), in particular:

> The following sentences illustrate the uses of *whose*:
> * ...
> * As the possessive of relative *which* (again, normally only as determiner): *It is an idea whose time has come* (alternatively *...of which the time has come*).

See also [this discussion on English Stack Exchange](https://english.stackexchange.com/questions/3965/which-whose-or-something-else), whose(!) answers unanimously support this and link to further sources.

**Suggestion:** Add `Whose` as an alias for `Which`. (Long-term you might want to deprecate `Which`, but I don't really care about that, I just want to use `Whose`.)

**Comment by dennisdoomen** [comment]
Interesting thought. I wonder what the native English speakers think of this...

**Comment by cmeeren** [comment]
AFAIK it's a fairly common mistake (like *their*/*they're* etc.), so you'd need to find someone at least a bit knowledgeable. Was the sources I linked to not satisfactory?

**Comment by cmeeren** [comment]
See e.g. examples on [Merriam-Webster](https://www.merriam-webster.com/dictionary/whose):

* the law courts, whose decisions were important
* the first poem whose publication he ever sanctioned

**Comment by adamvoss** [comment]
> Interesting thought. I wonder what the native English speakers think of this...

It depends the context.  _Whose_ is right when it is possessive.  When it is not possessive, _whose_ sounds more wrong that using _which_ in the possessive context.

Using the [examples](http://fluentassertions.com/examples.html) from the website it would come out as:

```c#
dictionary.Should().ContainValue(myClass).Whose.SomeProperty.Should().BeGreaterThan(0);
someObject.Should().BeOfType<Exception>().Whose.Message.Should().Be("Other Message");
xDocument.Should().HaveElement("child").Which.Should().BeOfType<XElement>().And.HaveAttribute("attr", "1");
```
Note the last example was unchanged.

**tl;dr** You can get better sounding "sentences" by including `Whose`, but both terms would need to be around forever, possibly creating API confusion.

**Comment by Im5tu** [comment]
The confusion is already there in some respects as `Which` and `Subject` are synonymous ([see here](https://github.com/fluentassertions/fluentassertions/blob/develop/Src/Core/AndWhichConstraint.cs#L60-L74)).

@adamvoss I suggest your last example should actually read:

    xDocument.Should().HaveElement("child").That.Should().BeOfType<XElement>().And.HaveAttribute("attr", "1");

**Comment by cmeeren** [comment]
@adamvoss I hadn't considered the third use case, you're right. Though as @Im5tu says, it should be `That` (see e.g. [here](http://www.quickanddirtytips.com/education/grammar/which-versus-that-0)).

In the general sense, one should of course not go overboard with synonyms. But in this particular case, I don't see the problem. The great thing about FluentAssertions is that you can write tests that read like English, and the `Which`/`Subject` property is a very simple thing to alias. By including `Whose`/`That`, a lot more sentences will read correctly instead of sounding rather off.

Personally I wasn't confused by the existing `Which`/`Subject`, because the documentation (Intellisense) clearly stated that they were synonyms. Heck, if you want, you can even make it even more clear by explicitly stating that the synonyms are there to facilitate more natural sentences.

**Comment by dennisdoomen** [comment]
So the conclusion of this issue is to add `Whose` and `That` anywhere where `Which` is available?
